configuration/config2/ConfigPath.cpp

```diff
@@ -0,0 +1,169 @@
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "ConfigPath.hpp"
+#include "Exceptions.hpp"
+#include "Utils.hpp"
+
+
+bool ConfigPathItem::isValuePresent(const std::string val, bool returnTrueIfValueListEmpty) const
+{
+    if (m_attributeValues.empty())
+    {
+        return returnTrueIfValueListEmpty;
+    }
+    else
+    {
+        for (auto it=m_attributeValues.begin(); it!=m_attributeValues.end(); ++it)
+        {
+            if (*it == val)
+                return true;
+        }
+    }
+    return false;
+}
+
+
+std::shared_ptr<ConfigPathItem> ConfigPath::getNextPathItem()
+{
+    std::shared_ptr<ConfigPathItem> pPathItem;
+
+    //
+    // If notheing left, return an empty pointer
+    if (m_path.empty())
+    {
+        return pPathItem;
+    }
+
+    //
+    // There is more to do, allocate a pointer and get to work
+    pPathItem = std::make_shared<ConfigPathItem>();
+
+    //
+    // Root
+    if (m_path[0] == '/')
+    {
+        parsePathElement(1, pPathItem);
+        pPathItem->setIsRoot(true);
+        updatePath(1);
+    }
+
+    //
+    // Parent (MUST be done before current element test)
+    else if (m_path.substr(0,2) == "..")
+    {
+        pPathItem->setIsParentElement(true);
+        updatePath(2);
+    }
+
+    //
+    // Current element?
+    else if (m_path[0] == '.')
+    {
+        pPathItem->setIsCurrentElement(true);
+        //m_path = m_path.substr(1);
+        updatePath(1);
+    }
+
+    //
+    // Otherwise, need to parse it out
+    else
+    {
+        parsePathElement(0, pPathItem);
+        updatePath(0);
+    }
+
+    return pPathItem;
+}
+
+
+void ConfigPath::parsePathElement(std::size_t start, const std::shared_ptr<ConfigPathItem> &pPathItem)
+{
+    std::size_t slashPos = m_path.find_first_of('/', start);
+    std::size_t len = (slashPos != std::string::npos) ? (slashPos-start) : std::string::npos;
+    std::string element = m_path.substr(start, len);
+
+    //
+    // The attribute definition is enclosed by the brackets, extract it if present
+    std::string attr;
+    if (extractEnclosedString(element, attr, '[', ']', true))
+    {
+        //
+        // Make sure valid
+        if (attr.empty())
+        {
+            throw(ParseException("Bad path, missng attribute definition at or around: " + element));
+        }
+
+        //
+        // The attribute must begin with a '@' or '#'. The '#' is an extension for config manager that allows the selection
```

Comment:
where will this notation be applied? is this an internal notation that only the core library knows about, or is this an expected notation in the XSDs? Also, is it compliant?

Reply:
would this '#' based syntax be applied in an XSD? if so, we definitely need to discuss. If it's an entirely internal concept to the core lib, it's perfectly fine.

Reply:
This is mainly used for internal processing, not for component XSDs.